pkg/apis/storage/types.go

```diff
@@ -424,3 +436,67 @@ type CSINodeList struct {
 	// items is the list of CSINode
 	Items []CSINode
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+//
+// This is an alpha feature and only available when the CSIStorageCapacity feature is enabled.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning. It must be
+	// be a DNS subdomain (dots allowed, 253 characters). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	// This field is immutable.
+	DriverName string
+
+	// NodeTopology defines which nodes have access to the storage for which
+	// capacity was reported. If not set, the storage is accessible from all
+	// nodes in the cluster. Only filtering by labels is supported.
+	// This field is immutable.
+	//
+	// +optional
+	NodeTopology *api.NodeSelector
+
+	// The name of the StorageClass that the reported capacity applies to.
+	// This field is immutable.
+	StorageClassName string
+
+	// Capacity is the value reported by the CSI driver in its GetCapacityResponse
+	// for a GetCapacityRequest with topology and parameters that match the
+	// previous fields.
+	//
+	// The semantic is currently (CSI spec 1.2) defined as:
+	// The available capacity, in bytes, of the storage that can be used
+	// to provision volumes. If not set, that information is currently
+	// unavailable and treated like zero capacity.
```

Comment:
Thanks for documenting the unset semantic.  What is the semantic for a zone or StorageClass that does not have a `CSIStorageCapacity` resource at all?

Reply:
That also means that this zone + storage class currently have no capacity left. This is important because the storage class parameters might select a certain pool that is simply not in the zone, and therefore the capacity would always be zero. Such combinations shouldn't have to be stored in etcd.

It's debatable whether a `CSIStorageCapacity` object with zero capacity then should be allowed at all. I think it makes sense because the semantic is clear and in the future we may want to allow objects where `Capacity` is unset or zero and some other fields are used instead.

Reply:
How does one enable this feature on an existing cluster?  If you enable it in the CSIDriver, there will be no capacity, so any volume requests will fail.  Do you have to pre-populate all the capacity resources before you enable the scheduler to consider capacity?

Are there any bootstrapping considerations for new clusters?

Reply:
Yeah it's a two step process. First the driver needs to be updated so that the capacity resources are created, and then you can flip the bit in CSIDriver to tell the scheduler to start looking at it.

Reply:
Is that well documented?  Eg. this enable field could say something about that :)

Reply:
It's in the KEP and it's probably also going to be in the end-user documentation, but I can also already add a comment here saying something about the intention.

Reply:
Done.